File: Avaliacoes/CSharp/Modulo2/controller/RelatorioController.cs
using System.Collections.Generic;

/// <summary>
/// Aplica o padrão Fork-Join:
/// dispara uma thread por filial (fork), aguarda todas (join) e soma os
/// 4 resultados finais (merge).
/// </summary>
public class RelatorioController
{
    private List<Filial> _filiais;
    private RelatorioView _view;
    private readonly List<FilialThread> _threads = new List<FilialThread>();

    public RelatorioController(List<Filial> filiais, RelatorioView view)
    {
        _filiais = filiais;
        _view = view;
    }

    public List<Filial> Filiais
    {
        get => _filiais;
        set => _filiais = value;
    }

    public RelatorioView View
    {
        get => _view;
        set => _view = value;
    }

    public void GerarRelatorio()
    {
        // FORK: uma thread por filial, cada uma com seus próprios dados.
        foreach (var f in _filiais)
        {
            _threads.Add(new FilialThread(f, _view));
        }

        foreach (var t in _threads)
        {
            t.Start();
        }

        // JOIN: aguarda todas terminarem antes de seguir.
        foreach (var t in _threads)
        {
            t.Join();
        }

        // MERGE: só lê os resultados depois que todas já terminaram.
        long faturamentoTotal = 0;
        foreach (var t in _threads)
        {
            faturamentoTotal += t.GetResultado().FaturamentoTotal;
        }

        _view.ExibirResultadoFinal(faturamentoTotal);
    }
}

File: Avaliacoes/Java/Modulo2/src/main/java/controller/RelatorioController.java
package controller;

import model.Filial;
import thread.FilialThread;
import view.RelatorioView;

import java.util.ArrayList;
import java.util.List;

/**
 * Responsável por aplicar o padrão Fork-Join: disparar uma thread por filial (fork), aguardar todas
 * (join) e somar os 4 resultados finais.
 *
 * @author Andrisa Santos
 */
public class RelatorioController {

    private List<Filial> filiais;
    private RelatorioView view;
    private List<FilialThread> threads;

    public RelatorioController(List<Filial> filiais, RelatorioView view) {
        this.filiais = filiais;
        this.view = view;
        this.threads = new ArrayList<>();
    }

    public List<Filial> getFiliais() {
        return filiais;
    }

    public void setFiliais(List<Filial> filiais) {
        this.filiais = filiais;
    }

    public RelatorioView getView() {
        return view;
    }

    public void setView(RelatorioView view) {
        this.view = view;
    }

    /**
     * @throws InterruptedException se a thread principal for interrompida
     * enquanto aguarda o término das demais
     */
    public void gerarRelatorio() throws InterruptedException {
        for (Filial f : filiais) {
            threads.add(new FilialThread(f, view));
        }

        for (FilialThread t : threads) t.start();   // fork
        for (FilialThread t : threads) t.join();    // join

        long faturamentoTotal = 0;
        for (FilialThread t : threads) {
            faturamentoTotal += t.getResultado().getFaturamentoTotal(); // merge, só após join
        }

        view.exibirResultadoFinal(faturamentoTotal);
    }
}
